### PR TITLE
fix: bound ABI array decode, correct out-of-gas detection, tighten AA-code match

### PR DESCRIPTION
**Pull request by Sednaoui**
Three correctness fixes found while reviewing #197, each reproduced against the branch code.

- `decodeAbiParameters`: a dynamic array's length word (untrusted RPC/bundler data) drove `Array(len).fill(...)` before any bounds check. A 64-byte payload claiming 2^27 elements OOM-crashes the process; ethers throws `BUFFER_OVERRUN` on the same input. Now bounded against the payload size before allocating.
- `decodeUserOperationRevertReason`: the EntryPoint emits `UserOperationRevertReason` only for non-empty revert data, so a failed op with no such log is the out-of-gas case. It was returning `outOfGas: false` there and `true` only for a fabricated empty-reason log real EntryPoints never emit. Branches swapped.
- `parseAaCode`: `/AA\d\d/i` matched digit pairs inside hex blobs (e.g. `0x3aa21f`, Sepolia chainId `0xaa36a7`). Now requires word boundaries.

Tests added for each; affected suites pass.

## Summary by CodeRabbit

* **Bug Fixes**
  * Prevented false error-code matches inside encoded blobs
  * Added bounds checks for dynamic-array decoding to avoid crashes on malformed payloads
  * Corrected handling of missing revert logs to mark out-of-gas cases

* **Tests**
  * Added/updated tests for error parsing, ABI decoding edge cases, and revert handling

* **Documentation**
  * Updated changelog with unreleased items including a new gas-estimation method and widened signature return types

### CODE REVIEW

File: test/userOperationRevert.test.js [hunk 1]
```diff
@@ -43,10 +43,13 @@ describe("decodeUserOperationRevertReason", () => {
...
 
-	it("reverted without a reason when there is no revert log", () => {
+	it("flags out-of-gas when a failed op has no revert log", () => {
+		// The EntryPoint emits UserOperationRevertReason only for non-empty
+		// revert data, so a failed op with no such log left no revert data:
+		// out-of-gas / bare revert() / call-to-non-contract.
 		const r = decodeUserOperationRevertReason({ success: false, logs: [], receipt: { logs: [] } });
 		expect(r.reverted).toBe(true);
-		expect(r.outOfGas).toBe(false);
+		expect(r.outOfGas).toBe(true);
```
[the remark this answers is not in the document]
Comment (Sednaoui) [reply test/userOperationRevert.test.js:52]: this is the repo's established convention

File: src/ethereUtils.ts [hunk 2]
```diff
@@ -826,6 +826,11 @@ function decodeValue(t: AbiType, data: Uint8Array, base: number): unknown {
             if (t.size === -1) {
                 ensureRange(data, base, WordSize, "array length");
                 const len = toNumber(data.slice(base, base + WordSize));
+                // Every element occupies at least one word (its value or its
+                // offset slot), so a length claiming more words than the buffer
+                // can hold is malformed. Bound it before allocating to avoid an
+                // out-of-memory crash on a hostile/short payload.
+                ensureRange(data, base + WordSize, len * WordSize, "array elements");
```
[the remark this answers is not in the document]
Comment (Sednaoui) [reply src/ethereUtils.ts]: fixed

**Review by sherifahmed990 — APPROVED** [review]
LGTM
